daffodil-runtime1-unparser/src/main/scala/edu/illinois/ncsa/daffodil/processors/unparsers/BCDUnparsers.scala

```diff
@@ -0,0 +1,135 @@
+
+  def unparse(state: UState): Unit = {
+    val nBits = getBitLength(state)
+    val node = state.currentInfosetNode.asSimple
+    val value = node.dataValue.asInstanceOf[JNumber]
+    val dos = state.dataOutputStream
+
+    val res =
+      if (nBits > 0) {
+        putNumber(dos, value, nBits)
+      } else {
+        true
+      }
+
+    if (!res) {
+      Assert.invariant(dos.maybeRelBitLimit0b.isDefined)
+      UnparseError(One(state.schemaFileLocation), One(state.currentLocation), "Insufficient space to unparse element %s, required %s bits, but only %s were available.",
+        e.dpathElementCompileInfo.namedQName.toPrettyString, nBits, dos.maybeRelBitLimit0b.get)
+    }
+  }
+
+}
+
+abstract class BCDIntegerBaseUnparser(e: ElementRuntimeData)
+  extends BCDNumberBaseUnparser(e) {
+
+  override def putNumber(dos: DataOutputStream, value: JNumber, nBits: Int): Boolean = {
+    dos.putByteArray(bcdFromBigInteger(new JBigInteger(value.toString), nBits), nBits)
+  }
+}
+
+class BCDIntegerKnownLengthUnparser(e: ElementRuntimeData, override val lengthInBits: Int)
+  extends BCDIntegerBaseUnparser(e)
+  with HasKnownLengthInBits {
+}
+
+class BCDIntegerRuntimeLengthUnparser(val e: ElementRuntimeData, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends BCDIntegerBaseUnparser(e)
+  with HasRuntimeExplicitLength {
+
+  override val runtimeDependencies = List(lengthEv)
+}
+
+class BCDDecimalKnownLengthUnparser(e: ElementRuntimeData, binaryDecimalVirtualPoint: Int, val lengthInBits: Int)
+  extends BCDDecimalBaseUnparser(e, binaryDecimalVirtualPoint)
+  with HasKnownLengthInBits {
+}
+
+class BCDDecimalRuntimeLengthUnparser(val e: ElementRuntimeData, binaryDecimalVirtualPoint: Int, val lengthEv: Evaluatable[JLong], val lUnits: LengthUnits)
+  extends BCDDecimalBaseUnparser(e, binaryDecimalVirtualPoint)
+  with HasRuntimeExplicitLength {
+
+  override val runtimeDependencies = List(lengthEv)
+}
+
+abstract class BCDDecimalBaseUnparser(e: ElementRuntimeData, binaryDecimalVirtualPoint: Int)
+  extends BCDNumberBaseUnparser(e) {
+
+  override def putNumber(dos: DataOutputStream, value: JNumber, nBits: Int): Boolean = {
+    dos.putByteArray(bcdFromBigDecimal(new JBigDecimal(value.toString), nBits), nBits)
+  }
+}
+
+final class BCDIntegerMinLengthInBytesUnparser(minLengthInBytes: Int, e: ElementRuntimeData)
+  extends BCDIntegerBaseUnparser(e) {
+
+  override def getBitLength(state: ParseOrUnparseState): Int = {
+    val len = state.currentNode.get.asSimple.dataValue.asInstanceOf[Array[Byte]].length * 8
+    val min = minLengthInBytes *  8
+    scala.math.max(len, min)
+  }
+}
+
+final class BCDDecimalMinLengthInBytesUnparser(minLengthInBytes: Int, e: ElementRuntimeData, binaryDecimalVirtualPoint: Int)
+  extends BCDDecimalBaseUnparser(e, binaryDecimalVirtualPoint) {
+
+  override def getBitLength(state: ParseOrUnparseState): Int = {
+    val len = state.currentNode.get.asSimple.dataValue.asInstanceOf[Array[Byte]].length * 8
```

Comment:
Same question here? Is this correct or an errant copy/paste? I don't see any tests for delimited hex binary, so I suspect this isn't thoroughly tested enough?